File: backend/app/api/impact.py
from fastapi import APIRouter, HTTPException, Query, status
from pydantic import BaseModel, Field
from typing import Dict, Any, Optional, List
import uuid

router = APIRouter(prefix="/api/impact", tags=["Metric Impact Simulator"])

# Pydantic Schemas for Strict JSON Contract (Section 20 & 26)
class DefinitionPayload(BaseModel):
    formula: str

class ProposedChangePayload(BaseModel):
    type: str = "formula_change"  # formula_change, filter_change, dimension_change, data_source_change, etc.
    formula: Optional[str] = None
    filters: Optional[str] = None
    dimension: Optional[str] = None
    data_source: Optional[str] = None

class SimulationScopePayload(BaseModel):
    region: Optional[str] = "Europe"
    period: Optional[str] = "Q3 2026"

class SimulationRequest(BaseModel):
    metric: str = "gross_margin"
    current_version: Optional[str] = "2.1"
    proposed_change: Optional[ProposedChangePayload] = None
    # Support alternate naming matching strict contract
    change_type: Optional[str] = None
    current_definition: Optional[DefinitionPayload] = None
    proposed_definition: Optional[DefinitionPayload] = None
    scope: Optional[SimulationScopePayload] = None
    simulation_scope: Optional[SimulationScopePayload] = None
    user_name: Optional[str] = "Rajesh Kapoor"

class ScenarioRequest(BaseModel):
    metric: Optional[str] = "gross_margin"
    formula: Optional[str] = "((Revenue - Cost - Logistics Cost) / Revenue) * 100"

# Approved semantic measures for governance checks (Section 18)
APPROVED_MEASURES = {
    "revenue", "cost", "logistics_cost", "logistics", "adjusted_cost",
    "gross_profit", "operating_expenses", "expenses", "order_count",
    "orders", "total_customers", "customers", "churned_customers"
}

# Catalog of metrics supported for simulation
METRICS_IMPACT_CATALOG = {
    "gross_margin": {
        "id": "gross_margin",
        "name": "gross_margin",
        "display_name": "Gross Margin",
        "owner": "Priya Sharma",
        "owner_role": "VP Strategic Finance",
        "current_version": "2.1",
        "status": "Verified",
        "formula": "((Revenue - Cost) / Revenue) * 100",
        "unit": "percentage",
        "current_value": 27.20,
        "dependencies_count": 17,
        "dependent_metrics_count": 3
    },
    "revenue": {
        "id": "revenue",
        "name": "revenue",
        "display_name": "Recognized Revenue",
        "owner": "Priya Sharma",
        "owner_role": "VP Strategic Finance",
        "current_version": "2.4",
        "status": "Verified",
        "formula": "SUM(revenue)",
        "unit": "currency",
        "current_value": 48.60,
        "dependencies_count": 24,
        "dependent_metrics_count": 4
    },
    "cost": {
        "id": "cost",
        "name": "cost",
        "display_name": "Cost of Goods Sold (COGS)",
        "owner": "Anand Verma",
        "owner_role": "Director of Financial Operations",
        "current_version": "2.1",
        "status": "Verified",
        "formula": "SUM(cost)",
        "unit": "currency",
        "current_value": 35.38,
        "dependencies_count": 15,
        "dependent_metrics_count": 3
    },
    "churn_rate": {
        "id": "churn_rate",
        "name": "churn_rate",
        "display_name": "Customer Churn Rate",
        "owner": "Vikram Malhotra",
        "owner_role": "Head of Customer Success",
        "current_version": "2.0",
        "status": "Verified",
        "formula": "(Churned Customers / Total Customers) * 100",
        "unit": "percentage",
        "current_value": 4.80,
        "dependencies_count": 9,
        "dependent_metrics_count": 2
    }
}

IN_MEMORY_SIMULATIONS = {
    "SIM-2026-00981": {
        "simulation_id": "SIM-2026-00981",
        "metric": "gross_margin",
        "current_version": "v2.1",
        "proposed_version": "Draft",
        "user": "Rajesh Kapoor",
        "current_formula": "((Revenue - Cost) / Revenue) * 100",
        "proposed_formula": "((Revenue - Cost - Logistics Cost) / Revenue) * 100",
        "current_value": 27.20,
        "simulated_value": 22.88,
        "difference": -4.32,
        "affected_assets": 17,
        "status": "Simulation Only"
    }
}

@router.get("/metrics")
async def get_impact_metrics():
    return {
        "status": "success",
        "count": len(METRICS_IMPACT_CATALOG),
        "metrics": list(METRICS_IMPACT_CATALOG.values())
    }

@router.get("/metric/{metric_id}/dependencies")
async def get_metric_dependencies(metric_id: str):
    metric = METRICS_IMPACT_CATALOG.get(metric_id)
    if not metric:
        raise HTTPException(status_code=404, detail=f"Metric '{metric_id}' not found in governed catalog.")
    
    return {
        "status": "success",
        "metric_id": metric["id"],
        "metric_name": metric["display_name"],
        "current_version": metric["current_version"],
        "affected_assets_count": 17,
        "dependent_metrics_count": 3,
        "dashboards_count": 5,
        "reports_count": 8,
        "saved_insights_count": 14,
        "saved_queries_count": 42
    }

@router.post("/simulate")
async def simulate_metric_impact(req: SimulationRequest):
    metric_id = req.metric.lower().strip()
    if metric_id not in METRICS_IMPACT_CATALOG:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Unknown metric '{metric_id}'. Simulation blocked."
        )

    # Determine proposed formula
    proposed_formula = ""
    if req.proposed_change and req.proposed_change.formula:
        proposed_formula = req.proposed_change.formula
    elif req.proposed_definition and req.proposed_definition.formula:
        proposed_formula = req.proposed_definition.formula
    else:
        proposed_formula = "((Revenue - Cost - Logistics Cost) / Revenue) * 100"

    # Governance Check: Unknown measure check (Section 18)
    # Check if unknown measure like 'shipping_cost_v2' is used
    raw_tokens = proposed_formula.replace("(", " ").replace(")", " ").replace("+", " ").replace("-", " ").replace("*", " ").replace("/", " ").split()
    for token in raw_tokens:
        clean = token.strip().lower()
        if clean in ["sum", "count", "avg", "nullif", "100", "0", "1"]:
            continue
        if clean.isdigit():
            continue
        # If token has alphabetic chars and is not approved
        if clean.isalpha() and clean not in APPROVED_MEASURES and not any(m in clean for m in APPROVED_MEASURES):
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail=f"Simulation Blocked. Unknown measure: '{clean}'. This measure is not registered in the Semantic Layer."
            )

    # Deterministic simulation computation
    current_value = 27.20
    simulated_value = 22.88
    if "logistics" in proposed_formula.lower():
        simulated_value = 22.88
    elif "adjusted" in proposed_formula.lower():
        simulated_value = 26.20
    else:
        simulated_value = 24.80

    difference = round(simulated_value - current_value, 2)
    sim_id = f"SIM-2026-{uuid.uuid4().hex[:5].upper()}"

    res = {
        "status": "success",
        "simulation_id": sim_id,
        "metric": metric_id,
        "current_value": current_value,
        "simulated_value": simulated_value,
        "difference": difference,
        "affected_assets": 17,
        "dependent_metrics": 3,
        "simulation_only": True,
        "is_demo_mode": True,
        "governance_status": "PASSED"
    }

    IN_MEMORY_SIMULATIONS[sim_id] = res
    return res

@router.get("/simulation/{simulation_id}")
async def get_simulation_by_id(simulation_id: str):
    if simulation_id not in IN_MEMORY_SIMULATIONS:
        raise HTTPException(status_code=404, detail=f"Simulation '{simulation_id}' not found.")
    return IN_MEMORY_SIMULATIONS[simulation_id]

@router.post("/scenarios")
async def get_what_if_scenarios(req: ScenarioRequest):
    return {
        "status": "success",
        "metric": req.metric,
        "scenarios": [
            {"id": "curr", "name": "Current Definition", "value": 27.2, "diff": 0.0},
            {"id": "a", "name": "Logistics Cost +5%", "value": 26.8, "diff": -0.4},
            {"id": "b", "name": "Logistics Cost +10%", "value": 26.3, "diff": -0.9},
            {"id": "c", "name": "Logistics Cost +20%", "value": 25.4, "diff": -1.8},
            {"id": "prop", "name": "Full Surcharge Deduction", "value": 22.88, "diff": -4.32}
        ]
    }

@router.get("/assets")
async def get_impact_assets(type: Optional[str] = Query(None)):
    return {
        "status": "success",
        "metric": "gross_margin",
        "dashboards_count": 5,
        "reports_count": 8,
        "saved_insights_count": 14,
        "queries_count": 42
    }

File: frontend/src/types/impact.ts
import { z } from "zod";

export type ChangeType =
  | "formula_change"
  | "filter_change"
  | "dimension_change"
  | "data_source_change"
  | "business_rule_change"
  | "time_logic_change";

export type ImpactLevel = "LOW" | "MEDIUM" | "HIGH";

export type AssetType =
  | "dashboard"
  | "report"
  | "saved_insight"
  | "query"
  | "dependent_metric"
  | "alert"
  | "api_consumer";

export interface GovernedMeasure {
  id: string;
  name: string;
  display_name: string;
  table: string;
  column: string;
  description: string;
  category: string;
  status: "approved" | "deprecated" | "draft";
}

export interface MetricImpactMetadata {
  id: string;
  name: string;
  display_name: string;
  owner: string;
  owner_role: string;
  current_version: string;
  status: "Verified" | "Under Review" | "Draft" | "Deprecated";
  formula: string;
  formula_display: string;
  formula_sql: string;
  category: string;
  unit: "percentage" | "currency" | "count" | "ratio";
  currency?: string;
  current_value: number;
  dependencies_count: number;
  dependent_metrics_count: number;
  components: {
    revenue?: number;
    cost?: number;
    logistics_cost?: number;
    operating_expenses?: number;
    customers?: number;
    churned_customers?: number;
    orders?: number;
  };
}

export interface ValidationCheckItem {
  id: string;
  name: string;
  passed: boolean;
  message: string;
  severity: "error" | "warning" | "success";
}

export interface GovernanceValidationResult {
  is_valid: boolean;
  blocked: boolean;
  block_reason?: string;
  unknown_measures: string[];
  unknown_dimensions: string[];
  checks: ValidationCheckItem[];
  timestamp: string;
}

export interface AffectedAssetItem {
  id: string;
  name: string;
  type: AssetType;
  owner: string;
  owner_role?: string;
  impact_reason: string;
  last_updated: string;
  widgets_count?: number;
  widgets?: string[];
  sections?: string[];
  query_sql?: string;
  execution_frequency?: string;
  severity: "high" | "medium" | "low";
}

export interface DependentMetricImpact {
  id: string;
  name: string;
  current_formula: string;
  proposed_formula: string;
  current_value: number;
  simulated_value: number;
  difference: number;
  unit: string;
  impact_path: string;
}

export interface ImpactAssessment {
  score: number; // 0 - 100
  level: ImpactLevel;
  total_affected_assets: number;
  dashboards_count: number;
  reports_count: number;
  saved_insights_count: number;
  queries_count: number;
  dependent_metrics_count: number;
  alerts_count: number;
  api_consumers_count: number;
  methodology: string;
  calculation_explanation: string[];
}

export interface WhatIfScenarioItem {
  id: string;
  name: string;
  description: string;
  parameter: string;
  parameter_delta: string;
  current_value: number;
  simulated_value: number;
  difference_pp: number;
  is_active: boolean;
}

export interface SimulationResult {
  simulation_id: string;
  metric_id: string;
  metric_name: string;
  current_version: string;
  proposed_version: string;
  change_type: ChangeType;
  current_definition: {
    formula: string;
    filters?: string;
    dimension?: string;
    data_source?: string;
  };
  proposed_definition: {
    formula: string;
    filters?: string;
    dimension?: string;
    data_source?: string;
  };
  scope: {
    region: string;
    period: string;
  };
  current_value: number;
  simulated_value: number;
  difference: number;
  difference_pp: number;
  unit: string;
  simulation_only: boolean;
  is_demo_mode: boolean;
  created_at: string;
  created_by: string;
  status: "DRAFT_SIMULATION" | "UNDER_REVIEW" | "APPROVED" | "BLOCKED";
  validation: GovernanceValidationResult;
  impact_assessment: ImpactAssessment;
  dependent_metrics: DependentMetricImpact[];
  affected_assets: AffectedAssetItem[];
  scenarios: WhatIfScenarioItem[];
  dependency_graph: {
    nodes: Array<{
      id: string;
      name: string;
      category: "metric" | "dashboard" | "report" | "insight" | "query" | "widget";
      level: number;
      impact: "direct" | "indirect" | "none";
      details: string;
    }>;
    links: Array<{
      source: string;
      target: string;
      label?: string;
    }>;
  };
}

export interface SimulationAuditRecord {
  simulation_id: string;
  metric_id: string;
  metric_name: string;
  current_version: string;
  proposed_version: string;
  user_name: string;
  user_role: string;
  timestamp: string;
  change_type: ChangeType;
  current_formula: string;
  proposed_formula: string;
  current_value: number;
  simulated_value: number;
  difference_pp: number;
  affected_assets_count: number;
  validation_status: "PASSED" | "FAILED" | "BLOCKED";
  status: "Simulation Only" | "Under Review" | "Approved";
}

// Zod schema for AI Agent strict JSON contract (Section 20)
export const StrictSimulationContractSchema = z.object({
  metric: z.string().min(1, "Metric identifier is required"),
  change_type: z.enum([
    "formula_change",
    "filter_change",
    "dimension_change",
    "data_source_change",
    "business_rule_change",
    "time_logic_change"
  ]),
  current_definition: z.object({
    formula: z.string().min(1, "Current formula is required")
  }),
  proposed_definition: z.object({
    formula: z.string().min(1, "Proposed formula is required")
  }),
  simulation_scope: z
    .object({
      region: z.string().optional(),
      period: z.string().optional()
    })
    .optional()
});

export type StrictSimulationContract = z.infer<typeof StrictSimulationContractSchema>;
